src/ucp/core/ucp_mm.c

```diff
@@ -105,37 +105,53 @@ ucs_status_t ucp_mem_rereg_mds(ucp_context_h context, ucp_md_map_t reg_md_map,
         } else if (!length) {
             /* don't register zero-length regions */
             continue;
-        } else if ((md_attr->cap.flags & UCT_MD_FLAG_REG) &&
-                   (md_attr->cap.reg_mem_types & UCS_BIT(mem_type))) {
-            ucs_assert(address && length);
+        } else if (md_attr->cap.flags & UCT_MD_FLAG_REG) {
+            if (!(md_attr->cap.reg_mem_types & UCS_BIT(mem_type))) {
+                status = UCS_ERR_UNSUPPORTED;
+            } else {
+                ucs_assert(address && length);
+
+                /* MD supports registration, register new memh on it */
+                status = uct_md_mem_reg(context->tl_mds[md_index].md, address,
+                        length, uct_flags, &uct_memh[memh_index]);
+            }
 
-            /* MD supports registration, register new memh on it */
-            status = uct_md_mem_reg(context->tl_mds[md_index].md, address,
-                                    length, uct_flags, &uct_memh[memh_index]);
-            if (status != UCS_OK) {
-                level = (uct_flags & UCT_MD_MEM_FLAG_HIDE_ERRORS) ?
-                        UCS_LOG_LEVEL_DEBUG : UCS_LOG_LEVEL_ERROR;
-                ucs_log(level,
-                        "failed to register address %p length %zu on md[%d]=%s: %s",
-                        address, length, md_index, context->tl_mds[md_index].rsc.md_name,
-                        ucs_status_string(status));
-                ucp_mem_rereg_mds(context, 0, NULL, 0, 0, alloc_md, mem_type,
-                                  alloc_md_memh_p, uct_memh, md_map_p);
-                return status;
+            if (status == UCS_OK) {
+                ucs_trace("registered address %p length %zu on md[%d] memh[%d]=%p",
+                        address, length, md_index, memh_index,
+                        uct_memh[memh_index]);
+                new_md_map |= UCS_BIT(md_index);
+                ++memh_index;
+                continue;
             }
 
-            ucs_trace("registered address %p length %zu on md[%d] memh[%d]=%p",
-                      address, length, md_index, memh_index,
-                      uct_memh[memh_index]);
-            new_md_map |= UCS_BIT(md_index);
-            ++memh_index;
+            level = (uct_flags & UCT_MD_MEM_FLAG_HIDE_ERRORS) ?
```

Comment:
maybe better to assign correct `level` at the beginning of the function, because you assign the same value on each iteration here, but value can't be changed

Reply:
no on each iteration, but just in case of error. I'd leave it as it is

Reply:
> no on each iteration, but just in case of error. I'd leave it as it is

yes, you're right - I didn't notice the `continue` operator in the `if (status == UCS_OK)` statement above. thank you 👍